receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go

```diff
@@ -0,0 +1,131 @@
+		if point.Interval.StartTime != nil && point.Interval.StartTime.IsValid() {
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(point.Interval.StartTime.AsTime()))
+		}
+
+		// Check if EndTime is set and valid
+		if point.Interval.EndTime != nil && point.Interval.EndTime.IsValid() {
+			dp.SetTimestamp(pcommon.NewTimestampFromTime(point.Interval.EndTime.AsTime()))
+		} else {
+			warnMsg := fmt.Sprintf("EndTime is invalid for metric: %s", ts.GetMetric().GetType())
+			mb.logger.Warn(warnMsg)
+			continue
+		}
+
+		switch v := point.Value.Value.(type) {
+		case *monitoringpb.TypedValue_DoubleValue:
+			dp.SetDoubleValue(v.DoubleValue)
+		case *monitoringpb.TypedValue_Int64Value:
+			dp.SetIntValue(v.Int64Value)
+		default:
+			log.Printf("Unhandled metric value type: %T", v)
+		}
+	}
+
+	return m
+}
+
+func (mb *MetricsBuilder) ConvertSumToMetrics(ts *monitoringpb.TimeSeries, m pmetric.Metric) pmetric.Metric {
+	m.SetName(ts.GetMetric().GetType())
+	m.SetUnit(ts.GetUnit())
+	sum := m.SetEmptySum()
+	sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+
+	for _, point := range ts.GetPoints() {
+		dp := sum.DataPoints().AppendEmpty()
+
+		// Directly check and set the StartTimestamp if valid
+		if point.Interval.StartTime != nil && point.Interval.StartTime.IsValid() {
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(point.Interval.StartTime.AsTime()))
+		}
+
+		// Check if EndTime is set and valid
+		if point.Interval.EndTime != nil && point.Interval.EndTime.IsValid() {
+			dp.SetTimestamp(pcommon.NewTimestampFromTime(point.Interval.EndTime.AsTime()))
+		} else {
+			warnMsg := fmt.Sprintf("EndTime is invalid for metric: %s", ts.GetMetric().GetType())
+			mb.logger.Warn(warnMsg)
+			continue
+		}
+
+		switch v := point.Value.Value.(type) {
+		case *monitoringpb.TypedValue_DoubleValue:
+			dp.SetDoubleValue(v.DoubleValue)
+		case *monitoringpb.TypedValue_Int64Value:
+			dp.SetIntValue(v.Int64Value)
+		default:
+			log.Printf("Unhandled metric value type: %T", v)
+		}
+	}
+
+	return m
+}
+
+func (mb *MetricsBuilder) ConvertDeltaToMetrics(ts *monitoringpb.TimeSeries, m pmetric.Metric) pmetric.Metric {
+	m.SetName(ts.GetMetric().GetType())
+	m.SetUnit(ts.GetUnit())
+	sum := m.SetEmptySum()
+	sum.SetAggregationTemporality(pmetric.AggregationTemporalityDelta)
+
+	for _, point := range ts.GetPoints() {
+		dp := sum.DataPoints().AppendEmpty()
+
+		// Directly check and set the StartTimestamp if valid
+		if point.Interval.StartTime != nil && point.Interval.StartTime.IsValid() {
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(point.Interval.StartTime.AsTime()))
+		}
+
+		// Check if EndTime is set and valid
+		if point.Interval.EndTime != nil && point.Interval.EndTime.IsValid() {
+			dp.SetTimestamp(pcommon.NewTimestampFromTime(point.Interval.EndTime.AsTime()))
+		} else {
+			warnMsg := fmt.Sprintf("EndTime is invalid for metric: %s", ts.GetMetric().GetType())
+			mb.logger.Warn(warnMsg)
+			continue
```

Comment:
Actually, should we delete the datapoint if it is invalid?  Otherwise we will be sending bad data here.  The same question applies to an invalid value type below

Reply:
I believe there's no need to delete the data point values since we're already skipping them, which ensures that no incorrect data will be sent further.

Reply:
The way the pmetric API works is that once you call AppendEmpty(), it has already been added to the batch.  Then, you just fill in details.  By continuing here, you are just leaving the metric as-is.

Reply:
Updated